File: leetcode/101-200/118-pascals-triangle/bench/generate.c
/* LeetCode #118 — C mirror, additive Pascal's triangle.
 * Same algorithm + workload as generate.kara: each rep builds a full triangle of a data-dependent
 * row count (30 + acc%16) as a NESTED `long**` with per-row malloc (structural parity with the
 * Vec[Vec] the others build — a flat contiguous array would give C an unfair cache edge), and folds
 * every entry. K = 80000. */
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007LL

/* Build a triangle: tri[i] is a malloc'd row of (i+1) longs. *rows_out set to num_rows. */
static long long **generate(long long num_rows) {
    long long **tri = (long long **)malloc(sizeof(long long *) * num_rows);
    for (long long i = 0; i < num_rows; i++) {
        long long *row = (long long *)malloc(sizeof(long long) * (i + 1));
        for (long long j = 0; j <= i; j++) {
            if (j == 0 || j == i) row[j] = 1;
            else row[j] = tri[i - 1][j - 1] + tri[i - 1][j];
        }
        tri[i] = row;
    }
    return tri;
}

static void free_tri(long long **tri, long long num_rows) {
    for (long long i = 0; i < num_rows; i++) free(tri[i]);
    free(tri);
}

static long long triangle_hash(long long **tri, long long num_rows) {
    long long h = 1;
    for (long long i = 0; i < num_rows; i++) {
        for (long long j = 0; j <= i; j++) h = (h * 131 + tri[i][j]) % MOD;
        h = (h * 31 + (i + 1) + 7) % MOD;
    }
    return h;
}

int main(void) {
    long long acc = 0;
    for (long long rep = 0; rep < 80000; rep++) {
        long long rows = 30 + (acc % 16);
        long long **tri = generate(rows);
        long long h = triangle_hash(tri, rows);
        acc = (acc * 131 + h) % MOD;
        free_tri(tri, rows);
    }
    printf("%lld\n", acc);
    return 0;
}

File: leetcode/101-200/118-pascals-triangle/bench/generate.rs
// LeetCode #118 — Rust mirror, additive Pascal's triangle.
// Same algorithm + workload as generate.kara: each rep builds a full triangle of a data-dependent
// row count (30 + acc%16) as a nested `Vec<Vec<i64>>` (matched to Kara's `Vec[Vec[i64]]`), folding
// every entry. K = 80000. Kara checks integer overflow by default, so the like-for-like row is
// `rustc -O -C overflow-checks=on`.

const MOD: i64 = 1000000007;

fn generate(num_rows: i64) -> Vec<Vec<i64>> {
    let mut tri: Vec<Vec<i64>> = Vec::new();
    for i in 0..num_rows {
        let mut row: Vec<i64> = Vec::new();
        for j in 0..=i {
            if j == 0 || j == i {
                row.push(1);
            } else {
                row.push(tri[(i - 1) as usize][(j - 1) as usize] + tri[(i - 1) as usize][j as usize]);
            }
        }
        tri.push(row);
    }
    tri
}

fn triangle_hash(tri: &[Vec<i64>]) -> i64 {
    let mut h: i64 = 1;
    for row in tri {
        for &x in row {
            h = (h * 131 + x) % MOD;
        }
        h = (h * 31 + row.len() as i64 + 7) % MOD;
    }
    h
}

fn main() {
    let mut acc: i64 = 0;
    for _ in 0..80000 {
        let rows = 30 + (acc % 16);
        let tri = generate(rows);
        let h = triangle_hash(&tri);
        acc = (acc * 131 + h) % MOD;
    }
    println!("{}", acc);
}
